src/service/position-observer-impl.js

```diff
@@ -92,12 +107,33 @@ class AbstractPositionObserver {
           Math.floor(Math.random() * LOW_FIDELITY_FRAME_COUNT) : 0,
       trigger(position) {
         const prePos = entry.position;
+
         if (prePos
             && layoutRectEquals(prePos.positionRect, position.positionRect)
             && layoutRectEquals(prePos.viewportRect, position.viewportRect)) {
           // position doesn't change, do nothing.
           return;
         }
+
+        // Calculate height currently in view
+        const docViewTop = position.viewportRect.top;
+        const docViewBottom = position.viewportRect.bottom;
+
+        const elemTop = position.positionRect.top;
+        const elemBottom = position.positionRect.bottom;
+
+        position.visibleHeight = 0;
+        if (elemTop <= docViewTop) {
+          position.visibleHeight = elemBottom - docViewTop;
+          position.relativePos = RelativePositions.TOP;
+        } else if (elemBottom >= docViewBottom) {
+          position.visibleHeight = docViewBottom - elemTop;
+          position.relativePos = RelativePositions.BOTTOM;
+        } else {
+          position.visibleHeight = elemBottom - elemTop;
+          position.relativePos = RelativePositions.INVIEW;
+        }
+
         if (layoutRectsOverlap(position.positionRect, position.viewportRect)) {
```

Comment:
Your logic covers the logic here. Please remove the check to `layoutRectsOverlap` then. 
Also please move the logic to `layout-rect.js`.